module/functions/azure/aad/Assert-AzureAdGroupMembership.Tests.ps1

```diff
@@ -0,0 +1,225 @@
+
+            Assert-MockCalled Get-AzureAdDirectoryObject -Times 0
+        }
+    }
+
+    Context "Multiple groups found" {
+
+        Mock Get-AzADGroup { $mockDuplicateGroups }
+        Mock Get-AzureAdDirectoryObject {}
+
+        It "should throw an exception" {
+            { Assert-AzureAdGroupMembership -Name "a-common-group-name" -RequiredMembers @("user@nowwhere.org") } | 
+                Should -Throw "Found multiple matching groups: ObjectId=$($mockDuplicateGroups[0].Id); ObjectId=$($mockDuplicateGroups[1].Id);"
+
+            Assert-MockCalled Get-AzureAdDirectoryObject -Times 0
+        }
+    }
+
+    Context "Required members are already in the group (one member)" {
+
+        Mock Get-AzADGroup { $mockGroup }
+        Mock Get-AzADGroupMember { $mockGroupMembers[0] }
+        Mock Get-AzureAdDirectoryObject { $mockGroupMembers[0] }
+        Mock Add-AzADGroupMember {}
+        Mock Remove-AzADGroupMember {}
+
+        Assert-AzureAdGroupMembership `
+            -Name $mockGroup.DisplayName `
+            -RequiredMembers @($mockGroupMembers[0].Id)
+
+        It "should not try to update the group" {
+            Assert-MockCalled Get-AzureAdDirectoryObject -Times 1
+            Assert-MockCalled Add-AzADGroupMember -Times 0
+            Assert-MockCalled Remove-AzADGroupMember -Times 0
+        }
+    }
+
+    Context "Required members are already in the group (multiple members)" {
+
+        Mock Get-AzADGroup { $mockGroup }
+        Mock Get-AzADGroupMember { $mockGroupMembers }
+        Mock Get-AzureAdDirectoryObject { $mockGroupMembers[0] } -ParameterFilter { $Criterion -eq $mockGroupMembers[0].Id }
+        Mock Get-AzureAdDirectoryObject { $mockGroupMembers[1] } -ParameterFilter { $Criterion -eq $mockGroupMembers[1].Id }
+        Mock Add-AzADGroupMember {}
+        Mock Remove-AzADGroupMember {}
+
+        Assert-AzureAdGroupMembership `
+            -Name $mockGroup.DisplayName `
+            -RequiredMembers ($mockGroupMembers | Select-Object -ExpandProperty Id)
+
+        It "should not try to update the group" {
+            Assert-MockCalled Get-AzureAdDirectoryObject -Times 2
+            Assert-MockCalled Add-AzADGroupMember -Times 0
+            Assert-MockCalled Remove-AzADGroupMember -Times 0
+        }
+    }
+
+    Context "Missing single member" {
+
+        Mock Get-AzADGroup { $mockGroup }
+        Mock Get-AzADGroupMember { $mockGroupMembers[0] }
+        Mock Get-AzureAdDirectoryObject { $mockGroupMembers[0] } -ParameterFilter { $Criterion -eq $mockGroupMembers[0].Id }
+        Mock Get-AzureAdDirectoryObject { $mockGroupMembers[1] } -ParameterFilter { $Criterion -eq $mockGroupMembers[1].Id }
+        Mock Add-AzADGroupMember {}
+        Mock Remove-AzADGroupMember {}
+
+        Assert-AzureAdGroupMembership `
+            -Name $mockGroup.DisplayName `
+            -RequiredMembers ($mockGroupMembers | Select-Object -ExpandProperty Id)
+
+        It "should add the missing member to the group" {
```

Comment:
This surprised me, because this is not the behaviour I would expect from anything called "Assert".

I expect assertions to mean "I'm expecting this condition to be true, and if it's not then something has gone wrong somewhere and we need to raise the alert." I absolutely don't expect assertions to attempt to change the state to match what I said I expected the state already to be.

I just found this thread: https://github.com/PowerShell/PowerShell/issues/11710

As far as I can tell, there seemed to be a broad agreement that assert means what I said, but quite a lot of people seemed to be happy enough to use the verb `Assert` in things that would go on to effect changes despite knowing full well that this isn't what `Assert` implies, which I found a bit horrifying. But the main disappointing outcome seemed to be that there isn't a standard agreed verb that means "make it so if it isn't already".

So apparently there's no good solution here. But is `Assert` the least bad option?

Reply:
Personally I would like to use the verb `Ensure` but that is not an approved one and using `Set` often collides with existing cmdlets.  The intent is for our `Assert-*` functions to operate on a 'desired state' basis rather than a 'do this'.

Least bad seems a reasonable description - we already have a bunch of such functions, so changing the naming convention now would be a breaking change (although we could setup aliases).

Reply:
`Ensure` was the verb that came to my mind before opening that thread you linked to. Having looked through the list of approved verbs, it does seem like `Assert` is the least bad option for now.

We also use `Assert` in this way in many other places, so I think if we were to change, then it should be made to be consistent everywhere.